##### Copyright 2025 Google LLC.

In [ ]:
#@title Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

<table class="tfo-notebook-buttons" align="left">
  <td>
    <a target="_blank" href="https://ai.google.dev/gemma/docs/embeddinggemma/inference-embeddinggemma-with-sentence-transformers"><img src="https://ai.google.dev/static/site-assets/images/docs/notebook-site-button.png" height="32" width="32" />View on ai.google.dev</a>
  </td>
  <td>
    <a target="_blank" href="https://colab.research.google.com/github/google/generative-ai-docs/blob/main/site/en/gemma/docs/embeddinggemma/inference-embeddinggemma-with-sentence-transformers.ipynb""><img src="https://www.tensorflow.org/images/colab_logo_32px.png" />Run in Google Colab</a>
  </td>
  <td>
    <a target="_blank" href="https://kaggle.com/kernels/welcome?src=https://github.com/google/generative-ai-docs/blob/main/site/en/gemma/docs/embeddinggemma/inference-embeddinggemma-with-sentence-transformers.ipynb"><img src="https://www.kaggle.com/static/images/logos/kaggle-logo-transparent-300.png" height="32" width="70"/>Run in Kaggle</a>
  </td>
  <td>
    <a target="_blank" href="https://console.cloud.google.com/vertex-ai/workbench/deploy-notebook?download_url=https://raw.githubusercontent.com/google/generative-ai-docs/main/site/en/gemma/docs/embeddinggemma/inference-embeddinggemma-with-sentence-transformers.ipynb"><img src="https://ai.google.dev/images/cloud-icon.svg" width="40" />Open in Vertex AI</a>
  </td>
  <td>
    <a target="_blank" href="https://github.com/google/generative-ai-docs/blob/main/site/en/gemma/docs/embeddinggemma/inference-embeddinggemma-with-sentence-transformers.ipynb"><img src="https://www.tensorflow.org/images/GitHub-Mark-32px.png" />View source on GitHub</a>
  </td>
</table>

# Gerando Embeddings com Sentence Transformers

O EmbeddingGemma é um modelo de embedding leve e aberto, projetado para recuperação rápida e de alta qualidade em dispositivos comuns, como celulares. Com apenas 308 milhões de parâmetros, ele é eficiente o suficiente para executar técnicas avançadas de IA, como a Geração Aumentada de Recuperação (RAG), diretamente em sua máquina local, sem necessidade de conexão com a internet.

## Configuração

Antes de começar este tutorial, conclua as seguintes etapas:

* Obtenha acesso ao Gemma fazendo login no [Hugging Face](https://huggingface.co/google/embeddinggemma-300M) e selecionando **Confirmar licença** para um modelo Gemma.

* Gere um [Token de Acesso](https://huggingface.co/docs/hub/en/security-tokens#how-to-manage-user-access-token) do Hugging Face e use-o para fazer login no Colab.

Este notebook será executado na CPU ou na GPU.

### Instalar pacotes Python

Instale as bibliotecas necessárias para executar o modelo EmbeddingGemma e gerar embeddings. O Sentence Transformers é um framework Python para embeddings de texto e imagem. Para mais informações, consulte a documentação do [Sentence Transformers](https://www.sbert.net/).

In [ ]:
!pip install -U sentence-transformers git+https://github.com/huggingface/transformers@v4.56.0-Embedding-Gemma-preview

  Cloning https://github.com/huggingface/transformers (to revision v4.56.0-Embedding-Gemma-preview) to /tmp/pip-req-build-0rwuodcf
  Running command git clone --filter=blob:none --quiet https://github.com/huggingface/transformers /tmp/pip-req-build-0rwuodcf
  Running command git checkout -q 60b68e304cf4b6569b0660a13b558b929d4b0e77
  Resolved https://github.com/huggingface/transformers to commit 60b68e304cf4b6569b0660a13b558b929d4b0e77
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
  Created wheel for transformers: filename=transformers-4.57.0.dev0-py3-none-any.whl size=12604658 sha256=c3aeefc542b7a0308e26013217d59899c255463bba20b767be99dd493159e6a6
  Stored in directory: /tmp/pip-ephem-wheel-cache-x90wd3nh/wheels/3a/21/76/c31899bac2cf601d3c74091b26a413bc3fb54770d5ccb5c924
Successfully built transformers
  Attempting uninstall: transformers
    Found existing installation: transformers 4.57.1
    Uni

After you have accepted the license, you need a valid Hugging Face Token to access the model.

In [ ]:
import os
import sys
from google.colab import userdata
os.environ["HF_TOKEN"] = userdata.get("HF_TOKEN")

### Carregar Modelo

Use as bibliotecas `sentence-transformers` para criar uma instância de uma classe de modelo com o EmbeddingGemma.

In [ ]:
import torch
from sentence_transformers import SentenceTransformer

device = "cuda" if torch.cuda.is_available() else "cpu"

model_id = "google/embeddinggemma-300M"
model = SentenceTransformer(model_id).to(device=device)

print(f"Device: {model.device}")
print(model)
print("Total number of parameters in the model:", sum([p.numel() for _, p in model.named_parameters()]))

modules.json:   0%|          | 0.00/573 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/997 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/18.7k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/58.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/1.49k [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.21G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/1.16M [00:00<?, ?B/s]

tokenizer.model:   0%|          | 0.00/4.69M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/33.4M [00:00<?, ?B/s]

added_tokens.json:   0%|          | 0.00/35.0 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/662 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/312 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/134 [00:00<?, ?B/s]

2_Dense/model.safetensors:   0%|          | 0.00/9.44M [00:00<?, ?B/s]

config.json:   0%|          | 0.00/134 [00:00<?, ?B/s]

3_Dense/model.safetensors:   0%|          | 0.00/9.44M [00:00<?, ?B/s]

Device: cpu
SentenceTransformer(
  (0): Transformer({'max_seq_length': 2048, 'do_lower_case': False, 'architecture': 'Gemma3TextModel'})
  (1): Pooling({'word_embedding_dimension': 768, 'pooling_mode_cls_token': False, 'pooling_mode_mean_tokens': True, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False, 'pooling_mode_weightedmean_tokens': False, 'pooling_mode_lasttoken': False, 'include_prompt': True})
  (2): Dense({'in_features': 768, 'out_features': 3072, 'bias': False, 'activation_function': 'torch.nn.modules.linear.Identity'})
  (3): Dense({'in_features': 3072, 'out_features': 768, 'bias': False, 'activation_function': 'torch.nn.modules.linear.Identity'})
  (4): Normalize()
)
Total number of parameters in the model: 307581696


## Gerando incorporação (embedding)

Uma incorporação é uma representação numérica de um texto, como uma palavra ou frase, que captura seu significado semântico. Essencialmente, é uma lista de números (um vetor) que permite aos computadores compreender as relações e o contexto das palavras.

Vamos ver como EmbeddingGemma processaria três palavras diferentes `["Apple", "Maça", "Smartphone"]`.

EmbeddingGemma foi treinada em grandes quantidades de texto e aprendeu as relações entre palavras e conceitos.

In [ ]:
words = ["Apple", "Maça", "Smartphone"]
embeddings = model.encode(words)
print(embeddings)
for idx, embedding in enumerate(embeddings):
  print(f"Embedding {idx+1} (shape): {embedding.shape}")

[[-1.9407766e-01  5.5318428e-03  3.2086000e-02 ... -7.7971727e-02
  -2.1451356e-02  1.4890656e-02]
 [-1.8719043e-01  6.5460871e-03  2.8649773e-02 ... -6.3437842e-02
  -1.8085590e-02 -2.6828877e-03]
 [-2.1170253e-01  5.7830874e-07  2.4128906e-02 ... -3.7906397e-02
  -2.6972989e-02  7.5545027e-03]]
Embedding 1 (shape): (768,)
Embedding 2 (shape): (768,)
Embedding 3 (shape): (768,)


O modelo gera um vetor numérico para cada frase. Os vetores reais são muito longos (768), mas para simplificar, são apresentados com algumas dimensões.

A chave não são os números individuais em si, mas **a distância entre os vetores**. Se representássemos esses vetores em um espaço multidimensional, os vetores para `maça` e `apple` estariam muito próximos um do outro. E o vetor para `smartphone` pode estar longe de `maça`, mas perto de `apple`.

## Determinando Similaridade

Nesta seção, usamos embeddings para determinar o quão sementicamente semelhantes são as diferentes sentenças. Aqui mostramos exemplos com pontuações de similaridade altas, médias e baixas.

Alta similaridade:
- Frase A: “O chef preparou uma refeição deliciosa para os convidados”.
- Frase B: “Um saboroso jantar foi preparado pelo chef para os visitantes”.
- Raciocínio: Ambas as frases descrevem o mesmo evento usando palavras e estruturas gramaticais diferentes (voz ativa vs. voz passiva). Eles transmitem o mesmo significado central.

Semelhança Média:
- Frase A: “Ela é especialista em aprendizado de máquina”.
- Frase B: “Ele tem um profundo interesse em inteligência artificial”.
- Raciocínio: As frases estão relacionadas porque o aprendizado de máquina é um subcampo da inteligência artificial. No entanto, eles falam sobre pessoas diferentes com diferentes níveis de envolvimento (especialista versus interesse).

Baixa similaridade:
- Frase A: "O tempo em Tóquio está ensolarado hoje."
- Frase B: “Preciso comprar mantimentos para a semana”.
- Raciocínio: As duas frases tratam de tópicos completamente não relacionados e não compartilham nenhuma sobreposição semântica.

In [ ]:
sentence_high = [
    "O chef preparou uma refeição deliciosa para os convidados.",
    "Um jantar saboroso foi preparado pelo chef para os visitantes."
]
sentence_medium = [
    "Ela é especialista em aprendizado de máquina.",
    "Ele tem um grande interesse em inteligência artificial."
]
sentence_low = [
    "O tempo em Tóquio está ensolarado hoje.",
    "Preciso comprar mantimentos para a semana."
]

for sentence in [sentence_high, sentence_medium, sentence_low]:
  print("🙋‍♂️")
  print(sentence)
  embeddings = model.encode(sentence)
  similarities = model.similarity(embeddings[0], embeddings[1])
  print("`-> 🤖 score: ", similarities.numpy()[0][0])

🙋‍♂️
['O chef preparou uma refeição deliciosa para os convidados.', 'Um jantar saboroso foi preparado pelo chef para os visitantes.']
`-> 🤖 score:  0.84263945
🙋‍♂️
['Ela é especialista em aprendizado de máquina.', 'Ele tem um grande interesse em inteligência artificial.']
`-> 🤖 score:  0.5167952
🙋‍♂️
['O tempo em Tóquio está ensolarado hoje.', 'Preciso comprar mantimentos para a semana.']
`-> 🤖 score:  0.2061559


In [ ]:
print("Available tasks:")
for name, prefix in model.prompts.items():
  print(f" {name}: \"{prefix}\"")
print("-"*80)

for sentence in [sentence_high, sentence_medium, sentence_low]:
  print("🙋‍♂️")
  print(sentence)
  embeddings = model.encode(sentence, prompt_name="STS")
  similarities = model.similarity(embeddings[0], embeddings[1])
  print("`-> 🤖 score: ", similarities.numpy()[0][0])


Available tasks:
 query: "task: search result | query: "
 document: "title: none | text: "
 BitextMining: "task: search result | query: "
 Clustering: "task: clustering | query: "
 Classification: "task: classification | query: "
 InstructionRetrieval: "task: code retrieval | query: "
 MultilabelClassification: "task: classification | query: "
 PairClassification: "task: sentence similarity | query: "
 Reranking: "task: search result | query: "
 Retrieval: "task: search result | query: "
 Retrieval-query: "task: search result | query: "
 Retrieval-document: "title: none | text: "
 STS: "task: sentence similarity | query: "
 Summarization: "task: summarization | query: "
--------------------------------------------------------------------------------
🙋‍♂️
['O chef preparou uma refeição deliciosa para os convidados.', 'Um jantar saboroso foi preparado pelo chef para os visitantes.']
`-> 🤖 score:  0.9433898
🙋‍♂️
['Ela é especialista em aprendizado de máquina.', 'Ele tem um grande interess

## Classificação

Classificar um texto é a tarefa de atribuí-lo a uma ou mais categorias ou rótulos predefinidos. É uma das tarefas mais fundamentais no Processamento de Linguagem Natural (PLN).

Uma aplicação prática da classificação de texto é o encaminhamento de chamados de suporte ao cliente. Esse processo direciona automaticamente as solicitações dos clientes para o departamento correto, economizando tempo e reduzindo o trabalho manual.

In [ ]:
labels = ["Problema de Cobrança", "Suporte Técnico", "Consulta de Vendas"]

sentence = [
"Com licença, o aplicativo trava na tela de login. Não funciona nem quando tento redefinir minha senha.",
"Gostaria de consultar os preços e recursos do plano empresarial para uma equipe de 50 pessoas.",
]

# Calculate embeddings by calling model.encode()
label_embeddings = model.encode(labels, prompt_name="Classification")
embeddings = model.encode(sentence, prompt_name="Classification")

# Calculate the embedding similarities
similarities = model.similarity(embeddings, label_embeddings)
print(similarities)

idx = similarities.argmax(1)
print(idx)

for example in sentence:
  print("🙋‍♂️", example, "-> 🤖", labels[idx[sentence.index(example)]])

tensor([[0.4686, 0.5009, 0.3581],
        [0.4838, 0.4481, 0.5243]])
tensor([1, 2])
🙋‍♂️ Com licença, o aplicativo trava na tela de login. Não funciona nem quando tento redefinir minha senha. -> 🤖 Suporte Técnico
🙋‍♂️ Gostaria de consultar os preços e recursos do plano empresarial para uma equipe de 50 pessoas. -> 🤖 Consulta de Vendas


# Referencia desse tutorial
https://ai.google.dev/gemma/docs/embeddinggemma/inference-embeddinggemma-with-sentence-transformers?hl=pt-br